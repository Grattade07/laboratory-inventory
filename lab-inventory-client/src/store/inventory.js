import {createSlice, current} from "@reduxjs/toolkit"

/* state slice for the application */
export const inventorySlice = createSlice({
    name: "inventory",

    /* initial set states */
    initialState: {
        inventoryUpdateCount: 0,
        inventoryList: [],
        displayedInventoryList: [],
        userJWToken: "",
        isAdmin: false,
        isLoggedIn: false
    },

    /* reducers to change state */
    reducers: {
        /* increments the update count */
        updateInventoryCount: (state) => {
            state.inventoryUpdateCount += 1
        },
        /* updates the inventory list */
        updateInventoryList: (state, action) => {
            state.inventoryList = action.payload
        },
        /* updates to the default inventory list to be shown in the table */
        displayList: (state) => {
            state.displayedInventoryList = state.inventoryList
        },
        /* makes the displayed table sorted by item name */
        displayItemSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.name > p2.name) ? 1 : (p1.name < p2.name) ? -1 : 0)
        },
        /* makes the displayed table sorted by item location */
        displayLocationSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.location > p2.location) ? 1 : (p1.location < p2.location) ? -1 : 0)
        },
        /* makes the displayed table sorted by item batch */
        displayBatchSortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.batch > p2.batch) ? 1 : (p1.batch < p2.batch) ? -1 : 0)
        },
        /* makes the displayed table sorted by item quantity */
        displayQuantitySortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.quantity > p2.quantity) ? 1 : (p1.quantity < p2.quantity) ? -1 : 0)
        },
        /* makes the displayed table sorted by item expiry */
        displayExpirySortedList: (state) => {
            state.displayedInventoryList = state.inventoryList.sort((p1, p2) => (p1.expiryDate > p2.expiryDate) ? 1 : (p1.expiryDate < p2.expiryDate) ? -1 : 0)
        },
        /* updates the stored JWToken */
        addJWToken: (state, action) => {
            state.userJWToken = action.payload
        },
        /* removes the stored JWToken */
        removeJWToken: (state) => {
            state.userJWToken = ""
        },
        /* changes the displayed table to what was search in the search bar */
        displayFilteredList: (state, action) => {
            let filterString = action.payload
            /* current function gets the current state of the inventoryList state. Found this function from url: https://redux-toolkit.js.org/api/other-exports#current */
            let inventoryList = current(state.inventoryList)
            let filteredList = function(string) {
                let filteredList = []

                for (let i = 0; i < inventoryList.length; i++) {
                    if(inventoryList[i].name.toLowerCase().includes(string)) {
                        filteredList.push(inventoryList[i])
                    }
                }

                return filteredList
            }

            state.displayedInventoryList = filteredList(filterString)
        },
        /* resets the displayed list to the original data */
        resetDisplayList: (state) => {
            state.displayedInventoryList = state.inventoryList
        },
        /* updates if user is admin */
        setIsAdmin: (state, action) => {
            state.isAdmin = action.payload
        },
        /* updates if user is logged in */
        setIsLoggedIn: (state, action) => {
            state.isLoggedIn = action.payload
        }
    }
})

export const {
    updateInventoryCount, 
    updateInventoryList, 
    displayList, 
    displayItemSortedList, 
    displayLocationSortedList, 
    displayBatchSortedList, 
    displayQuantitySortedList, 
    displayExpirySortedList, 
    addJWToken, 
    removeJWToken, 
    displayFilteredList, 
    resetDisplayList,
    setIsAdmin,
    setIsLoggedIn
} = inventorySlice.actions

export default inventorySlice.reducer