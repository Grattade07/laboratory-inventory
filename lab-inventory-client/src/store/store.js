import { configureStore } from "@reduxjs/toolkit";
import inventoryReducer from "./inventory"

/* store for application */
export default configureStore({
    reducer: {
        inventory : inventoryReducer
    }
})