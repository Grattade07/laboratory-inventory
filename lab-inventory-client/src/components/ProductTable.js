import Table from 'react-bootstrap/Table'
import { useEffect, useState } from "react";
import Button from "react-bootstrap/Button";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome"
import {faPlus, faMinus, faXmark} from "@fortawesome/free-solid-svg-icons"
import { useDispatch, useSelector } from 'react-redux';
import { updateInventoryList, displayList, updateInventoryCount } from '../store/inventory';

/* component for the table that will display the  */
function ProductTable () {
    const [isLoaded, setIsLoaded] = useState(false)
    const [error, setError] = useState(null)

    const dispatch = useDispatch()

    /* state selectors */

    const inventory = useSelector((state) => state.inventory.displayedInventoryList)

    const updatedCount = useSelector((state) => state.inventory.inventoryUpdateCount)

    const token = useSelector((state) => state.inventory.userJWToken)

    const isLoggedIn = useSelector((state) => state.inventory.isLoggedIn)

    const isAdmin = useSelector((state) => state.inventory.isAdmin)

    /* function sends a get reqeust to receive the inventory list from the database */
    function componentDidMount() {
        fetch("/inventory/getInventory")
          .then(res => res.json())
          .then(
            (result) => {
            setIsLoaded(true)
            dispatch(updateInventoryList(result))
            dispatch(displayList())
          },
          (error) => {
            setIsLoaded(true)
            setError(error)
          } ) 
        }
    
    /* function sends a put request to increment the value of a given item */
    function increaseQuantity(url ="", data = {}) {
        return fetch(url, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json",
                "token" : token
            },
            body: JSON.stringify(data)
        })
        .then(res => res.json())
        .then(res => dispatch(updateInventoryCount()))
    }

    /* function sends a put request to decrement the value of a given item */
    function decreaseQuantity(url ="", data = {}) {
        return fetch(url, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json",
                "token" : token
            },
            body: JSON.stringify(data)
        })
        .then(res => res.json())
        .then(res => dispatch(updateInventoryCount()))
    }

    /* function sends a delete request to remove an item from the database */
    function deleteItem(url ="", data ={}) {
        return fetch(url, {
            method: "DELETE",
            headers: {
                "Content-Type" : "application/json",
                "token" : token
            },
            body: JSON.stringify(data)
        })
        .then(res => res.json())
        .then(res => dispatch(updateInventoryCount()))
    }

    /* componentDidMount function runs on page load and when "updatedCount" changes */
    useEffect(() => componentDidMount(), [updatedCount])

    /* shows if there was an error getting the data */
    if (error) {
        return (
        <div id="table-container">
            <Table id="product-table" bordered striped>
                <thead id="table-header">
                    <tr>
                        <th>Item</th>
                        <th>Location</th>
                        <th>Batch</th>
                        <th>Quantity</th>
                        <th>Expiry Date</th>
                        <th>Purchase Link</th>
                    </tr>
                </thead>

                <tbody style={{height: "auto"}}>
                    <tr><td>Error: {error}</td></tr>
                </tbody>
            </Table>
        </div>
        )
    } 
    /* shows if the user is not logged in */
    else if (!isLoggedIn) {
        return (
            <div id="table-container">
                <Table id="product-table" bordered striped>
                    <thead id="table-header">
                        <tr>
                            <th>Item</th>
                            <th>Location</th>
                            <th>Batch</th>
                            <th>Quantity</th>
                            <th>Expiry Date</th>
                            <th>Purchase Link</th>
                        </tr>
                    </thead>
    
                    <tbody style={{height: "100%"}}>
                        <tr><td colSpan={6} rowSpan={11}>Please Login to see inventory...</td></tr>
                    </tbody>
                </Table>
            </div>
            )
    } 
    /* shows if the table has not loaded yet */
    else if(!isLoaded) {
        return (
        <div id="table-container">
            <Table id="product-table" bordered striped>
                <thead id="table-header">
                    <tr>
                        <th>Item</th>
                        <th>Location</th>
                        <th>Batch</th>
                        <th>Quantity</th>
                        <th>Expiry Date</th>
                        <th>Purchase Link</th>
                    </tr>
                </thead>

                <tbody style={{height: "auto"}}>
                    <tr><td colSpan={6} rowSpan={11}>Getting Items...</td></tr>
                </tbody>
            </Table>
        </div>
        )
    } 
    /* shows the inventory table */
    else {
    return (
        <div id="table-container">
            <Table id="product-table" bordered striped>
                <thead id="table-header">
                    <tr>
                        <th>Item</th>
                        <th>Location</th>
                        <th>Batch</th>
                        <th>Quantity</th>
                        <th>Expiry Date</th>
                        <th>Purchase Link</th>
                        <th>Remove</th>
                    </tr>
                </thead>

                <tbody>
                    {inventory.map(item => (
                        <tr key={item._id}>

                            <td>{item.name}</td>

                            <td>{item.location}</td>

                            <td>{item.batch}</td>
                            
                            <td className='item-quantity-display'>
                                {/* button sends request to increase item quantity. Is disabled if user is not an admin */}
                                <Button className='quantity-button' disabled={isAdmin ? false : true} onClick={() => {
                                    increaseQuantity("/inventory/increaseQuantity", {
                                        id: item._id,
                                        quantity: item.quantity + 1
                                    })
                                }
                                }>
                                    <FontAwesomeIcon icon={faPlus} className="quantity-button-icon"></FontAwesomeIcon>
                                </Button>

                                <div className="item-quantity">{item.quantity}</div>

                                {/* button sends request to decrease item quantity. Is disabled if user is not an admin */}
                                <Button className='quantity-button' disabled={isAdmin ? false : true} onClick={() => 
                                    decreaseQuantity("/inventory/decreaseQuantity", {
                                        id: item._id,
                                        quantity: item.quantity - 1
                                    })
                                }>
                                    <FontAwesomeIcon icon={faMinus} className="quantity-button-icon">
                                    </FontAwesomeIcon>
                                </Button>
                            </td>

                            <td>{item.expiryDate}</td>

                            {/* href opens link in new tab */}
                            <td>
                                <Button href={`https://${item.url}`} target="_blank" disabled={isAdmin ? false : true}>Purchase</Button>
                            </td>

                            <td>
                                {/* button sends request to delete the item. Is disabled if user is not an admin */}
                                <Button variant="danger" onClick={() => 
                                    deleteItem("/inventory/deleteItem", {
                                        id: item._id
                                    })
                                } disabled={isAdmin ? false : true}>
                                    <FontAwesomeIcon icon={faXmark} ></FontAwesomeIcon>
                                </Button>
                            </td>
                        </tr>
                    ))}
                </tbody>
            </Table>
        </div>
    )
    }
}

export default ProductTable