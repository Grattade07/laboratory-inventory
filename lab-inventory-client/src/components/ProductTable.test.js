import renderer from "react-test-renderer"
import ProductTable from "./ProductTable"
import { Provider } from 'react-redux';
import store from '../store/store';

/* tests that the filter component renders correctly  */
it("renders correctly", () => {
    const tree = renderer
        .create(<Provider store={store}>
                    <ProductTable />
                </Provider>
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})