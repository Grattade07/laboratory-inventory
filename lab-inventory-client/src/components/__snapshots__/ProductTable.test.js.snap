// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="table-container"
>
  <table
    className="table table-striped table-bordered"
    id="product-table"
  >
    <thead
      id="table-header"
    >
      <tr>
        <th>
          Item
        </th>
        <th>
          Location
        </th>
        <th>
          Batch
        </th>
        <th>
          Quantity
        </th>
        <th>
          Expiry Date
        </th>
        <th>
          Purchase Link
        </th>
      </tr>
    </thead>
    <tbody
      style={
        Object {
          "height": "100%",
        }
      }
    >
      <tr>
        <td
          colSpan={6}
          rowSpan={11}
        >
          Please Login to see inventory...
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
